### PR TITLE
UPGRADE jest & react-native

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: -jest is outdated and this is causing bithound to nag us-

## What I did
- [x] upgraded Jest
- [x] upgraded react-native
- [x] fixed tests / issues on core
- [x] fixed tests / issues on RN
- [x] fixed issues in RN examples
- [x] test React-Native examples

### CODE REVIEW

**Review by Hypnosphi — APPROVED** [review]

File: examples/cra-kitchen-sink/src/storyshots.test.js [hunk 1]
```diff
@@ -1,6 +1,14 @@
 import initStoryshots, { snapshotWithOptions } from '@storybook/addon-storyshots';
 import path from 'path';
 
+jest.mock('global', () => ({
```
Comment (ndelangen) [inline examples/cra-kitchen-sink/src/storyshots.test.js]: I think this should actually go in the storyshots addon code

File: examples/crna-kitchen-sink/package.json [hunk 2]
```diff
@@ -30,9 +30,9 @@
...
   },
   "dependencies": {
-    "expo": "^22.0.0",
-    "prop-types": "^15.5.0",
-    "react": "~16.0.0",
-    "react-native": "~0.49.3"
+    "expo": "^24.0.2",
+    "prop-types": "^15.6.0",
+    "react": "^16.2.0",
+    "react-native": "^0.51.0"
```
Comment (ndelangen) [inline examples/crna-kitchen-sink/package.json]: These upgrade worry me, need to be careful and test this

File: examples/react-native-vanilla/jest.config.js [hunk 3]
```diff
@@ -1,5 +1,12 @@
 module.exports = {
   preset: 'react-native',
+  clearMocks: true,
```
Comment (ndelangen) [inline examples/react-native-vanilla/jest.config.js]: Somewhere a png was being required into RN..

File: jest.config.js [hunk 4]
```diff
@@ -1,5 +1,5 @@
 module.exports = {
-  cacheDirectory: '.cache/jest',
+  // cacheDirectory: '.cache/jest',
```
Comment (ndelangen) [inline jest.config.js]: Might have to revert this

File: app/react/src/client/preview/index.js [hunk 5]
```diff
@@ -13,11 +13,12 @@ import { selectStory } from './actions';
...
 
 // check whether we're running on node/browser
-const { navigator } = global;
 const isBrowser =
   navigator &&
+  navigator.userAgent &&
   navigator.userAgent !== 'storyshots' &&
-  !(navigator.userAgent.indexOf('Node.js') > -1);
+  !(navigator.userAgent.indexOf('Node.js') > -1) &&
+  !(navigator.userAgent.indexOf('jsdom') > -1);
```
Comment (ndelangen) [inline app/react/src/client/preview/index.js]: This is the main difference with Jest22, it depends on a higher version of jsdom, and is has a new userAgent

File: lib/ui/src/modules/ui/components/search_box.stories.js [hunk 6]
```diff
@@ -17,7 +16,7 @@ const onClose = action('onClose');
 
 // For some reason react-modal causes an segfault (infinite loop maybe?)
 // when rendered by storyshots/react-test-renderer
-if (!navigator.userAgent.match(/Node\.js/)) {
+if (!navigator.userAgent.match(/Node\.js/) && !navigator.userAgent.match(/jsdom/)) {
```
Comment (ndelangen) [inline lib/ui/src/modules/ui/components/search_box.stories.js]: I kept both for compatibility

File: package.json [hunk 7]
```diff
@@ -62,14 +62,18 @@
     "glob": "^7.1.2",
     "husky": "^0.14.3",
     "inquirer": "^4.0.1",
-    "jest": "^21.2.0",
-    "jest-cli": "^21.2.1",
+    "jest": "^22.0.4",
```
Comment (ndelangen) [inline package.json]: I added these dependencies to make sure the right versions were hoisted to the top level.. It seems CRA has some old version of something

**Comment by ndelangen** [comment]
This should actually fix the bithound status, I'm confused why it's still red

**Comment by ndelangen** [comment]
I've tested RN on iOS, it seems to work just fine

**Comment by ndelangen** [comment]
Can someone else @storybooks/team please verify the RN stuff on their machine? 🙇

**Pull request merged**